From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Mon, 5 Feb 2018 22:13:09 +0100
Subject: [PATCH] {NOT UPSTREAMABLE} netfilter: silence -Warray-bounds

ARM gcc complains:

      CC [M]  net/bridge/br_netfilter_hooks.o
    In file included from include/linux/kernel.h:10:0,
                     from include/linux/list.h:9,
                     from include/linux/module.h:9,
                     from net/bridge/br_netfilter_hooks.c:17:
    net/bridge/br_netfilter_hooks.c: In function 'br_nf_post_routing':
    include/linux/compiler.h:291:20: error: array subscript is above
    array bounds [-Werror=array-bounds]
       __read_once_size(&(x), __u.__c, sizeof(x));  \
                        ^~~~
    include/linux/compiler.h:297:22: note: in expansion of macro '__READ_ONCE'
     #define READ_ONCE(x) __READ_ONCE(x, 1)
                          ^~~~~~~~~~~
    include/linux/rcupdate.h:349:48: note: in expansion of macro 'READ_ONCE'
      typeof(*p) *________p1 = (typeof(*p) *__force)READ_ONCE(p); \
                                                    ^~~~~~~~~
    include/linux/rcupdate.h:486:2: note: in expansion of macro '__rcu_dereference_check'
      __rcu_dereference_check((p), (c) || rcu_read_lock_held(), __rcu)
      ^~~~~~~~~~~~~~~~~~~~~~~
    include/linux/rcupdate.h:544:28: note: in expansion of macro 'rcu_dereference_check'
     #define rcu_dereference(p) rcu_dereference_check(p, 0)
                                ^~~~~~~~~~~~~~~~~~~~~
    include/linux/netfilter.h:219:15: note: in expansion of macro 'rcu_dereference'
       hook_head = rcu_dereference(net->nf.hooks_arp[hook]);
                   ^~~~~~~~~~~~~~~

Because nf_hooks() is called with a variable protocol family but a fixed
hook index which is over NF_ARP_NUMHOOKS. Ad a BUG_ON to make gcc happy.
---
 include/linux/netfilter.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/netfilter.h b/include/linux/netfilter.h
index dd2052f0efb7..a3ccffb08dcf 100644
--- a/include/linux/netfilter.h
+++ b/include/linux/netfilter.h
@@ -215,6 +215,8 @@ static inline int nf_hook(u_int8_t pf, unsigned int hook, struct net *net,
 		break;
 	case NFPROTO_ARP:
 #ifdef CONFIG_NETFILTER_FAMILY_ARP
+		/* When using ARP, ensure hook does not overflow */
+		BUG_ON(hook >= NF_ARP_NUMHOOKS);
 		hook_head = rcu_dereference(net->nf.hooks_arp[hook]);
 #endif
 		break;
-- 
